### PR TITLE
Add TryParse support for route parameters.

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR adds TryParse support to route parameters. Uses the same test data as RDF. Currently waiting for this PR to be approved then I'll rebase on main to solve an outstanding indentation issue.

https://github.com/dotnet/aspnetcore/pull/46928

### CODE REVIEW

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs [hunk 1]
```diff
@@ -76,7 +76,8 @@ internal static void EmitRouteParameterPreparation(this EndpointParameter endpoi
             codeWriter.EndBlock();
         }
 
-        codeWriter.WriteLine($"var {endpointParameter.EmitHandlerArgument()} = {endpointParameter.EmitAssigningCodeResult()};");
+        codeWriter.WriteLine($"var {endpointParameter.Name}_temp = {endpointParameter.EmitAssigningCodeResult()}?.ToString();");
```
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:79]:
```suggestion
        codeWriter.WriteLine($"var {endpointParameter.EmitTempArgument()} = {endpointParameter.EmitAssigningCodeResult()}?.ToString();");
```

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 2]
```diff
@@ -94,31 +96,31 @@ private bool TryGetParsability(IParameterSymbol parameter, WellKnownTypes wellKn
...
+            preferredTryParseInvocation = (string inputArgument, string outputArgument) => $$"""{{parameterType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}}.TryParse({{inputArgument}}!, CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal | DateTimeStyles.AllowWhiteSpaces, out var {{outputArgument}})""";
         }
         else if (SymbolEqualityComparer.Default.Equals(parameterType, wellKnownTypes.Get(WellKnownType.System_DateTimeOffset)))
         {
-            preferredTryParseInvocation = (string inputArgument, string outputArgument) => $$"""{{parameterType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}}.TryParse({{inputArgument}}, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal | DateTimeStyles.AllowWhiteSpaces, out var {{outputArgument}})""";
+            preferredTryParseInvocation = (string inputArgument, string outputArgument) => $$"""{{parameterType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}}.TryParse({{inputArgument}}!, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal | DateTimeStyles.AllowWhiteSpaces, out var {{outputArgument}})""";
```
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:119]: Can you explain the need for the null suppression operator here and elsewhere?

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs [hunk 3]
```diff
@@ -76,7 +76,8 @@ internal static void EmitRouteParameterPreparation(this EndpointParameter endpoi
...
+        EmitParsingBlock(endpointParameter, codeWriter);
```
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:80]:
```suggestion
        endpointParameter.EmitParsingBlock(codeWriter);
```

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs (hunk 2 above)
Comment (mitchdenny) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:119]: It probably isn't ideal. This is the situation that led to it:

```csharp
Task RequestHandler(HttpContext httpContext)
{
    var wasParamCheckFailure = false;
    var context_local = httpContext;
    // Endpoint Parameter: routeValue (Type = System.Net.IPEndPoint, IsOptional = False, IsParsable = True, Source = Route)
    if (options?.RouteParameterNames?.Contains("routeValue", StringComparer.OrdinalIgnoreCase) != true)
    {
        throw new InvalidOperationException($"'routeValue' is not a route parameter.");
    }
    var routeValue_raw = httpContext.Request.RouteValues["routeValue"]?.ToString();
    if (routeValue_raw == null)
    {
        wasParamCheckFailure = true;
    }
    var routeValue_temp = routeValue_raw?.ToString();
    if (!global::System.Net.IPEndPoint.TryParse(routeValue_temp!, out var routeValue_parsed_temp))
    ...
```

We check that ```RouteParameterNames``` contains ```routeValue``` and if it doesn't we throw. But then we get the value out of the collection, but because the compiler can't guarantee that its not null I used the ?.ToString() approach. So from the compilers perspective ```routeValue_raw``` and ```routeValue_temp``` could both be null.

I could adjust the ```if``` block containing ```TryParse``` to do a null check to avoid the usage of the null suppression but I'm not sure if adds much value. I guess I am assuming that ```TryParse``` implementations will return false when ```s``` is null. The fact that TryParse implementations are typically in the form ```TryParse(string? s, ....)``` makes me think that is a fair guess.

Basically, all those ```TryParse``` invocations could be dealing with a null input at some point.

**Review by captainsafia — APPROVED** [review]
